src/basic/chase-symlinks.c

```diff
@@ -400,6 +394,88 @@ int chase_symlinks(
         return 0;
 }
 
+int chase_symlinks(
+                const char *path,
+                const char *original_root,
+                ChaseSymlinksFlags flags,
+                char **ret_path,
+                int *ret_fd) {
+
+        _cleanup_free_ char *root = NULL, *absolute = NULL, *p = NULL;
+        _cleanup_close_ int fd = -1, pfd = -1;
+        int r;
+
+        assert(path);
+
+        if (isempty(path))
+                return -EINVAL;
+
+        /* A root directory of "/" or "" is identical to none */
+        if (empty_or_root(original_root))
+                original_root = NULL;
+
+        if (original_root) {
+                r = path_make_absolute_cwd(original_root, &root);
+                if (r < 0)
+                        return r;
+
+                /* Simplify the root directory, so that it has no duplicate slashes and nothing at the
+                 * end. While we won't resolve the root path we still simplify it. Note that dropping the
+                 * trailing slash should not change behaviour, since when opening it we specify O_DIRECTORY
+                 * anyway. Moreover at the end of this function after processing everything we'll always turn
+                 * the empty string back to "/". */
+                delete_trailing_chars(root, "/");
+                path_simplify(root);
+
+                if (flags & CHASE_PREFIX_ROOT) {
+                        absolute = path_join(root, path);
+                        if (!absolute)
+                                return -ENOMEM;
+                }
+        }
+
+        if (!absolute) {
+                r = path_make_absolute_cwd(path, &absolute);
+                if (r < 0)
+                        return r;
+        }
+
+        if (root) {
+                path = path_startswith(absolute, root);
+                if (!path)
+                        return log_full_errno(flags & CHASE_WARN ? LOG_WARNING : LOG_DEBUG,
+                                              SYNTHETIC_ERRNO(ECHRNG),
+                                              "Specified path '%s' is outside of specified root directory '%s', refusing to resolve.",
+                                              absolute, root);
+        }
+
+        /* chase_symlinks_at() disallows empty paths, so if the "path" is identical to "root", let's
+         * explicitly specify that we want to open the root directory by adding ".". */
+        if (isempty(path))
+                path = ".";
```

Comment:
wouldn't it make more sense to pass `path` as NULL in that case and let chase_symlinks_at() figure it out